tests/integration/acp.go

```diff
@@ -69,12 +113,301 @@ func addPolicyACP(
 	}
 }
 
-func getIdentity(s *state, index immutable.Option[int]) immutable.Option[acpIdentity.Identity] {
+func setupSourceHub(s *state) ([]node.ACPOpt, error) {
+	var isACPTest bool
+	for _, a := range s.testCase.Actions {
+		if _, ok := a.(AddPolicy); ok {
+			isACPTest = true
+			break
+		}
+	}
+	if !isACPTest {
+		// Spinning up SourceHub instances is a bit slow, so we should be quite aggressive in trimming down the
+		// runtime of the test suite when SourceHub ACP is selected.
+		s.t.Skipf("test has no ACP elements when testing with SourceHub ACP")
+	}
+
+	const moniker string = "foo"
+	const chainID string = "sourcehub-test"
+	const validatorName string = "test-validator"
+	const keyringBackend string = "test"
+	directory := s.t.TempDir()
+
+	kr, err := keyring.OpenFileKeyring(
+		directory,
+		keyring.PromptFunc(func(s string) ([]byte, error) {
+			return []byte("secret"), nil
+		}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate the keys using the index as the seed so that multiple
+	// runs yield the same private key.  This is important for stuff like
+	// the change detector.
+	source := rand.NewSource(0)
+	r := rand.New(source)
+
+	acpKey, err := secp256k1.GeneratePrivateKeyFromRand(r)
+	require.NoError(s.t, err)
+	acpKeyHex := hex.EncodeToString(acpKey.Serialize())
+
+	err = kr.Set(validatorName, acpKey.Serialize())
+	if err != nil {
+		return nil, err
+	}
+
+	out, err := exec.Command("sourcehubd", "init", moniker, "--chain-id", chainID, "--home", directory).CombinedOutput()
+	s.t.Log(string(out))
+	if err != nil {
+		return nil, err
+	}
+
+	// Annoyingly, the CLI does not support support changing the comet config params that we need,
+	// so we have to manually rewrite the config file.
+	cfg, err := toml.LoadFile(filepath.Join(directory, "config", "config.toml"))
+	if err != nil {
+		return nil, err
+	}
+
+	fo, err := os.Create(filepath.Join(directory, "config", "config.toml"))
+	if err != nil {
+		return nil, err
+	}
+
+	// Speed up the rate at which the blocks are created, this is particularly important for getting
+	// the first block created on the `sourcehubd start` call at the end of this function as
+	// we cannot use the node until the first block has been created.
+	cfg.Set("consensus.timeout_propose", "0.5s")
+	cfg.Set("consensus.timeout_commit", "1s")
```

Comment:
I don't think breaking up the setup CLI commands would help readability.  There are 200 literal lines, however not that many calls, and very little logic (the logic that exists is largely broken out to other funcs).  Fragmenting it would make it harder to read IMO.

Reply:
>  You, as the author of code, might find it easy to navigate through it, but for me, as a reader, it's hard.

This is a general thing, not specific to this function.  IMO a long flat function is much easier to read than a bunch of tiny ones.

In order to 'read' the main func in the fragment approach you have to trust that the names and documentation are good, and there are minimal side effects. And that is basically never the case, because the funcs are so small and specific the code within them ends up being as small and concise as the func signature/documentation in the first place - so instead of just reading what it does you have to bounce around either reading overly lengthy and potentially incorrect documentation, or you have to read the implementation anyway.

Apart from from a one-line for loop at the top of the func, the timeout stuff, and some error handling (required/doubled anyway with the fragment approach), there arent any branches/control-flow-statements in this func.  It is about as simple as it can get.

I assume you have seen plenty of previous code-horrors in your career caused by long funcs, I have seen a few too, however in my experience problems only really start to exist if the complexity, not line-count, is large.  And in my past, I have found tiny fragmented funcs to be a much larger problem.  We have different experiences here, and we both have valid, general opinions.  This func can remain as is, just as I avoid asking you to collapse your tiny funcs into a single func that I find easier to read.